integration/gethnetwork/geth_network.go

```diff
@@ -0,0 +1,172 @@
+	tempDirPrefix   = "geth_nodes"
+
+	addPeerCmd = "admin.addPeer(%s)"
+	attachCmd  = "attach"
+	enodeCmd   = "admin.nodeInfo.enode"
+	initCmd    = "init"
+
+	dataDirFlag      = "--datadir"
+	execFlag         = "--exec"
+	portFlag         = "--port"
+	mineFlag         = "--mine"
+	minerThreadsFlag = "--miner.threads=1"
+	minerEthbaseFlag = "--miner.etherbase=0x0000000000000000000000000000000000000001"
+
+	// We pre-allocate a single wallet, which is shared by all nodes.
+	genesisConfig = `{
+	  "config": {
+		"chainId": 777,
+		"homesteadBlock": 0,
+		"eip150Block": 0,
+		"eip155Block": 0,
+		"eip158Block": 0,
+		"byzantiumBlock": 0,
+		"constantinopleBlock": 0,
+		"petersburgBlock": 0,
+		"istanbulBlock": 0,
+		"berlinBlock": 0,
+		"londonBlock": 0
+	  },
+	  "alloc": {
+		"0x0000000000000000000000000000000000000001": {
+		  "balance": "111111111"
+		}
+	  },
+	  "coinbase": "0x0000000000000000000000000000000000000000",
+	  "difficulty": "0x20000",
+	  "extraData": "",
+	  "gasLimit": "0x2fefd8",
+	  "nonce": "0x0000000000000042",
+	  "mixhash": "0x0000000000000000000000000000000000000000000000000000000000000000",
+	  "parentHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
+	  "timestamp": "0x00"
+  }`
+)
+
+// GethNetwork is a network of Geth nodes, built using the provided Geth binary.
+type GethNetwork struct {
+	gethBinaryPath  string
+	genesisFilePath string
+	dataDirs        []string
+}
+
+// NewGethNetwork using the provided Geth binary to create a private Ethereum network with numNodes Geth nodes.
+func NewGethNetwork(gethBinaryPath string, numNodes int) GethNetwork {
+	nodesDir, err := ioutil.TempDir("", tempDirPrefix)
+	if err != nil {
+		panic(err)
+	}
+
+	// We write out the `genesis.json` file to be used by the network.
+	genesisFilePath := nodesDir + string(os.PathSeparator) + genesisFileName
+	err = os.WriteFile(genesisFilePath, []byte(genesisConfig), 0o600)
+	if err != nil {
+		panic(err)
+	}
+
+	// Each Geth node needs its own data directory.
+	dataDirs := make([]string, numNodes)
+	for i := 0; i < numNodes; i++ {
+		dataDirs[i] = fmt.Sprintf(nodesDir+nodeFolderName, i+1)
+	}
+
+	network := GethNetwork{
+		gethBinaryPath:  gethBinaryPath,
+		genesisFilePath: genesisFilePath,
+		dataDirs:        dataDirs,
+	}
+
+	for i, dataDir := range dataDirs {
+		_ = os.Remove(dataDir + ipcFileName) // We delete leftover IPC files from previous runs.
+		go network.createMiner(dataDir, 30303+i)
+	}
+
+	// We need to manually tell the nodes about one another.
+	network.joinNodesToNetwork(dataDirs)
+
+	return network
+}
+
+// IssueCommand sends the command via RPC to the nodeIdx'th node in the network.
+func (network *GethNetwork) IssueCommand(nodeIdx int, command string) string {
+	dataDir := network.dataDirs[nodeIdx]
+
+	args := []string{dataDirFlag, dataDir, attachCmd, dataDir + ipcFileName, execFlag, command}
+	cmd := exec.Command(network.gethBinaryPath, args...) // nolint
+
+	output, err := cmd.Output()
```

Comment:
What if these two methods just called the IssueCommand ?

Reply:
I don't think it would work. `IssueCommand` sends an `attach` command to the node, which is followed by a string command. For initialising and starting a node, we need to send a series of flags, etc.

There is also the issue that for `IssueCommand`, we want to wait for the output. For starting the node, we want to run the command but not wait for it to finish (as it spins forever).